.github/CONTRIBUTING.md

```diff
@@ -1,9 +1,73 @@
 # Contributing to Trino
 
-## Contributor License Agreement ("CLA")
 
-In order to accept your pull request, we need you to [submit a CLA](https://github.com/trinodb/cla).
+This is the process we suggest for contributions.  This process is designed to
+reduce the burden on project reviewers, impact on other contributors, and to
+keep the amount of rework from the contributor to a minimum.
 
-## License
+By contributing to Trino, you agree that your contributions will be licensed
+under the [Apache License Version 2.0 (APLv2)](../LICENSE).
 
-By contributing to Trino, you agree that your contributions will be licensed under the [Apache License Version 2.0 (APLv2)](../LICENSE).
+Find more information at [the Trino site](https://trino.io/development/).
+
+## Contributor License Agreement
+
+In order to accept your pull request, we need you to [submit a contributor
+license agreement (CLA)](https://github.com/trinodb/cla).
+
+## Discussion
+
+Start a discussion by creating a Github
+[issue](https://github.com/trinodb/trino/issues), or asking on
+[Slack](https://trino.io/slack.html) (unless the change is trivial).
+
+* This step helps you identify possible collaborators and reviewers.
+* Does the change align with technical vision and project values?
+* Will the change conflict with another change in progress? If so, work with
+  others to minimize impact.
+* Is this change large?  If so, work with others to break into smaller steps.
+
+## Implementation
+
+Implement the change.
+
+* If the change is large, post a preview Github
+  [pull request](https://github.com/trinodb/trino/pulls) with the title
+  prefixed with `[WIP]`, and share with collaborators.
+* Include tests and documentation as necessary.
+
+
+## Pull request
+
+Create a Github [pull request](https://github.com/trinodb/trino/pulls) (PR).
+
+* Make sure the pull request passes the tests in CI.
+* If known, request a review from an expert in the area changed.  If unknown,
+  ask for help on [Slack](https://trino.io/slack.html).
+
+## Review
+
+Review is performed by one or more reviewers.
+
+* This normally happens within a few days, but may take longer if the change is
+  large, complex, or if a critical reviewer is unavailable. Feel free to ping
+  the reviewer on the pull request.
+
+## Updates
+
+Address concerns and update the pull request.
+
+* Comments are addressed to each individual commit in the pull request, and
+  changes should be addressed in a new
+  [`fixup!` commit](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt)
+  placed after each commit. This is to make it easier for the reviewer to see
+  what was updated.
```

Comment:
I very much like this suggestion. We are not doing that typically - but we should - it makes reviews much easier.

Reply:
mostly maintainers and repeat contributors do that - others often don't. I've found it useful to explicitly ask for it when leaving review comment + a link to https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html

Reply:
> mostly maintainers and repeat contributors do that

good point.

i would expect experienced contributors not to read this document at all
and new contributors are typically not capable of doing pro git work, so don't expect too much

Reply:
> don't expect too much

To expect something, one must first communicate the requirements. That's why we need this guide. I'd like to suggest pulling these files up, to the root of the project, so they're easier to find.